### PR TITLE
Lower BatchNormalization (test mode) to Krnl dialect

### PR DESCRIPTION
**Pull request by tungld**
This patch is to lower BatchNormalization to Krnl dialect. It only supports the test mode of BatchNormalization.

Currently, e2e tests for batch normalization, including `test_batchnorm_epsilon_cpu` and `test_batchnorm_example_cpu` got segmentation fault. It is unknown to me why it happened, although ONNF correctly generated MLIR programs for them.

To verify that our lowering for batch normalization is correct, I did manual tests and the lowering worked well. Below are information.

An input program, `test_batchnorm.mlir`,  was the one got from running `test_batchnorm_example_cpu` (`onnx.EntryPoint` was commented since we didn't use it in this manual test):
```mlir
module {
  func @batchnorm_testmode_Nd(%arg0: tensor<1x2x1x3xf32>, %arg1: tensor<2xf32>, %arg2: tensor<2xf32>, %arg3: tensor<2xf32>, %arg4: tensor<2xf32>) -> tensor<1x2x1x3xf32> {
    %0 = "onnx.BatchNormalizationTestMode"(%arg0, %arg1, %arg2, %arg3, %arg4) : (tensor<1x2x1x3xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>) -> tensor<1x2x1x3xf32>
    return %0 : tensor<1x2x1x3xf32>
  }
   // "onnx.EntryPoint"() {func = @main_graph, numInputs = 5 : i32, numOutputs = 1 : i32} : () -> ()
}
```

A main C++ program, `batchnorm.cpp`, to call function `batchnorm_testmode_Nd`:
```c++
#include <iostream>
#include <vector>

using namespace std;

template <int N> void dropFront(int64_t arr[N], int64_t *res) {
  for (unsigned i = 1; i < N; ++i)
    *(res + i - 1) = arr[i];
}

/// StridedMemRef descriptor type with static rank.
template <typename T, int N> struct StridedMemRefType {
  T *basePtr;
  T *data;
  int64_t offset;
  int64_t sizes[N];
  int64_t strides[N];
  // This operator[] is extremely slow and only for sugaring purposes.
  StridedMemRefType<T, N - 1> operator[](int64_t idx) {
    StridedMemRefType<T, N - 1> res;
    res.basePtr = basePtr;
    res.data = data;
    res.offset = offset + idx * strides[0];
    dropFront<N>(sizes, res.sizes);
    dropFront<N>(strides, res.strides);
    return res;
  }
};

/// StridedMemRef descriptor type specialized for rank 1.
template <typename T> struct StridedMemRefType<T, 1> {
  T *basePtr;
  T *data;
  int64_t offset;
  int64_t sizes[1];
  int64_t strides[1];
  T &operator[](int64_t idx) { return *(data + offset + idx * strides[0]); }
};

/// StridedMemRef descriptor type specialized for rank 0.
template <typename T> struct StridedMemRefType<T, 0> {
  T *basePtr;
  T *data;
  int64_t offset;
};

extern "C" StridedMemRefType<float, 4> batchnorm_testmode_Nd(
    StridedMemRefType<float, 4> *a, StridedMemRefType<float, 1> *scale,
    StridedMemRefType<float, 1> *bias, StridedMemRefType<float, 1> *mean,
    StridedMemRefType<float, 1> *variance);

int main() {
  StridedMemRefType<float, 4> x; 
  StridedMemRefType<float, 1> scale, bias, mean, variance;

  // x :: 1x2x1x3
  float* arrX = (float*) malloc(1 * 2 * 1 * 3 * sizeof(float));
  x.data = arrX;
  x.basePtr = arrX;
  x.offset = 0;
  x.sizes[0] = 1;
  x.sizes[1] = 2;
  x.sizes[2] = 1;
  x.sizes[3] = 3;
  x.strides[0] = 6;
  x.strides[1] = 3;
  x.strides[2] = 3;
  x.strides[3] = 1;

  *(x.data + x.offset + 0 * x.strides[0] + 0 * x.strides[1] + 0 * x.strides[2] + 0 * x.strides[3]) = (float)-1;
  *(x.data + x.offset + 0 * x.strides[0] + 0 * x.strides[1] + 0 * x.strides[2] + 1 * x.strides[3]) = (float)0;
  *(x.data + x.offset + 0 * x.strides[0] + 0 * x.strides[1] + 0 * x.strides[2] + 2 * x.strides[3]) = (float)1;
  *(x.data + x.offset + 0 * x.strides[0] + 1 * x.strides[1] + 0 * x.strides[2] + 0 * x.strides[3]) = (float)2;
  *(x.data + x.offset + 0 * x.strides[0] + 1 * x.strides[1] + 0 * x.strides[2] + 1 * x.strides[3]) = (float)3;
  *(x.data + x.offset + 0 * x.strides[0] + 1 * x.strides[1] + 0 * x.strides[2] + 2 * x.strides[3]) = (float)4;

  // scale : 2
  float* arrScale = (float*) malloc(2 * sizeof(float));
  scale.data = arrScale;
  scale.basePtr = arrScale;
  scale.offset = 0;
  scale.sizes[0] = 2;
  scale.strides[0] = 1;
  *(scale.data + scale.offset + 0 * scale.strides[0]) = (float)1;
  *(scale.data + scale.offset + 1 * scale.strides[0]) = (float)1.5;

  // bias : 2
  float* arrBias = (float*) malloc(2 * sizeof(float));
  bias.data = arrBias;
  bias.basePtr = arrBias;
  bias.offset = 0;
  bias.sizes[0] = 2;
  bias.strides[0] = 1;
  *(bias.data + bias.offset + 0 * bias.strides[0]) = (float)0;
  *(bias.data + bias.offset + 1 * bias.strides[0]) = (float)1;

  // mean : 2
  float* arrMean = (float*) malloc(2 * sizeof(float));
  mean.data = arrMean;
  mean.basePtr = arrMean;
  mean.offset = 0;
  mean.sizes[0] = 2;
  mean.strides[0] = 1;
  *(mean.data + mean.offset + 0 * mean.strides[0]) = (float)0;
  *(mean.data + mean.offset + 1 * mean.strides[0]) = (float)3;

  // variance : 2
  float* arrVariance = (float*) malloc(2 * sizeof(float));
  variance.data = arrVariance;
  variance.basePtr = arrVariance;
  variance.offset = 0;
  variance.sizes[0] = 2;
  variance.strides[0] = 1;
  *(variance.data + variance.offset + 0 * variance.strides[0]) = (float)1;
  *(variance.data + variance.offset + 1 * variance.strides[0]) = (float)1.5;

  auto c = batchnorm_testmode_Nd(&x, &scale, &bias, &mean, &variance);

  // Print results.
  int I = 1, J = 2, K = 1, L =3;
  cout << "[";
  for (int i = 0; i < I; i++) {
    cout << "[";
    for (int j = 0; j < J; j++) {
      cout << "[";
      for (int k = 0; k < K; k++) {
        cout << "[";
        for (int l = 0; l < L; l++) {
          cout << c[i][j][k][l] << ((L > 1 && l < L - 1) ? ", " : "");
        }
        if (K > 1 && k < K - 1)
          cout << "], " << endl;
        else
          cout << "]";
      }
      if (J > 1 && j < J - 1)
        cout << "], " << endl;
      else
        cout << "]";
    } 
    if (I > 1 && i < I - 1)
      cout << "], " << endl;
    else
      cout << "]";
  }
  cout << "]" << endl;

  return 0;
}
```

Compile and run:
```bash
$ onnf --EmitLLVMBC test_batchnorm.mlir
$ llc --filetype=obj model.bc
$ g++ batchnorm.cpp model.o
$ ./a.out
[[[[-0.999995, 0, 0.999995]],
[[-0.224741, 1, 2.22474]]]]
```

Now, we check the above result again a result produced by NumPy for batch normalization defined in https://github.com/onnx/onnx/blob/master/onnx/backend/test/case/node/batchnorm.py:
```python
>>> import numpy as np
>>> x = np.array([[[[-1, 0, 1]], [[2, 3, 4]]]]).astype(np.float32)
>>> s = np.array([1.0, 1.5]).astype(np.float32)
>>> bias = np.array([0, 1]).astype(np.float32)
>>> mean = np.array([0, 3]).astype(np.float32)
>>> var = np.array([1, 1.5]).astype(np.float32)

>>> dims_x = len(x.shape)
>>> dim_ones = (1,) * (dims_x - 2)
>>> s = s.reshape(-1, *dim_ones)
>>> bias = bias.reshape(-1, *dim_ones)
>>> mean = mean.reshape(-1, *dim_ones)
>>> var = var.reshape(-1, *dim_ones)
>>> epsilon=1e-5
>>> y=s * (x - mean) / np.sqrt(var + epsilon) + bias
>>> y
array([[[[-0.99999499,  0.        ,  0.99999499]],
        [[-0.22474074,  1.        ,  2.22474074]]]], dtype=float32)
```

So, our lowering produced the same result as the result in numpy, and there was no segmentation fault.

### CODE REVIEW

**Comment by tjingrant** [comment]
I'm looking at how to address the segfault.

**Comment by tjingrant** [comment]
@tungld the fix is small, but I can't edit your branch directly. So please apply the following diff when you get a chance, thanks! I tested locally the batch norm tests now run correctly.

```diff
diff --git a/src/runtime/dyn_memref.cpp b/src/runtime/dyn_memref.cpp
index a20001c..454947b 100644
--- a/src/runtime/dyn_memref.cpp
+++ b/src/runtime/dyn_memref.cpp
@@ -35,7 +35,7 @@ DynMemRef *getDynMemRef(OrderedDynMemRefDict *tensorDict, int idx) {

 void setDynMemRef(OrderedDynMemRefDict *tensorDict, int idx,
                   DynMemRef *tensor) {
-  if (tensorDict->orderedNames.capacity() <= idx)
+  if (tensorDict->orderedNames.size() <= idx)
     tensorDict->orderedNames.resize(idx + 1);

   // The dynamic memref is essentially anonymous, since we are storing it by
```

**Comment by tungld** [comment]
Thanks, @tjingrant! it works perfectly.
